scipy/signal/fir_filter_design.py

```diff
@@ -583,3 +583,84 @@ def remez(numtaps, bands, desired, weight=None, Hz=1, type='bandpass',
     bands = np.asarray(bands).copy()
     return sigtools._remez(numtaps, bands, desired, weight, tnum, Hz,
                            maxiter, grid_density)
+
+
+def firls(numtaps, bands, desired, weight=None, Hz=1):
+    """
+    FIR filter design using least-squares error minimization.
+
+    Calculate the filter coefficients for the finite impulse response
+    (FIR) filter which has the best approximation to the desired
+    frequency response described by `bands` and `desired` in the
+    least squares sense.
+
+    Parameters
+    ----------
+    numtaps : int
+        The number of taps in the FIR filter.  `numtaps` must be odd.
+    bands : array_like
+        A monotonic sequence containing the band edges in Hz.
+        All elements must be non-negative and less than half the sampling
+        frequency as given by `Hz`.
+    desired : array_like
+        A sequence half the size of bands containing the desired gain
+        in each of the specified bands.
+    weight : array_like, optional
+        A relative weighting to give to each band region. The length of
+        `weight` has to be half the length of `bands`.
+    Hz : scalar, optional
+        The sampling frequency in Hz. Default is 1.
+
+    Returns
+    -------
+    out : ndarray
+        A rank-1 array containing the coefficients of the optimal
+        (in a least squares sense) filter.
+
+    Example
+    -------
+    We want to construct a filter with a passband at 0.2-0.3 Hz, and
+    stop bands at 0-0.1 Hz and 0.4-0.5 Hz. Note that this means that the
+    behavior in the frequency ranges between those bands is unspecified and
+    may overshoot.
+    >>> from scipy import signal
+    >>> bpass = signal.firls(71, [0, 0.1, 0.2, 0.3, 0.4, 0.5], [0, 1, 0])
+    >>> freq, response = signal.freqz(bpass)
+    >>> ampl = np.abs(response)
+    >>> import matplotlib.pyplot as plt
+    >>> fig = plt.figure()
+    >>> ax1 = fig.add_subplot(111)
+    >>> ax1.semilogy(freq/(2*np.pi), ampl, 'b-')  # freq in Hz
+    >>> plt.show()
+    """
+    if numtaps % 2 == 0:
+        raise ValueError("numtaps must be odd.")
+    L = (numtaps-1)//2
+
+    # normalize bands and make it 2 columns
+    bands = np.asarray(bands).flatten()/Hz
+    if len(bands) % 2 == 1:
+        raise ValueError("bands must contain frequency pairs.")
+    bands = bands.reshape(-1,2)
+
+    # check remaining params
+    if len(bands) != len(desired):
+        raise ValueError("desired must have one entry per band.")
+    if weight is None:
+        weight = np.ones_like(desired)
+
+    # set up the linear matrix equation to be solved, Ax = b
+    k = np.arange(L+1)[np.newaxis]
+    m = k.T
+    A,b = 0,0
+    for i, (f0,f1) in enumerate(bands):
+        Ai = f1 * (sinc(2*(m+k)*f1) + sinc(2*(m-k)*f1)) \
+           - f0 * (sinc(2*(m+k)*f0) + sinc(2*(m-k)*f0))
+        bi = desired[i] * (2*f1*sinc(2*m*f1) - 2*f0*sinc(2*m*f0))
+        A += Ai * abs(weight[i]**2)
```

Comment:
The weights should not be squared.  I tried comparing the output of this `firls` to Octave's `firls`, and I get agreement if the weights are not squared:

**`firls` from this pull request, but with the weights *not* squared:**

```
In [4]: taps = firls(9, [0, 0.5, 0.55, 1], [1, 0], [1, 2], Hz=2)

In [5]: taps
Out[5]: 
array([-0.00062693, -0.10335445, -0.00981577,  0.31727169,  0.51140943,
        0.31727169, -0.00981577, -0.10335445, -0.00062693])
```

**`firls` from the Octave signal package:**

```
octave:30> taps = firls(8, [0 0.5 0.55 1], [1 1 0 0], [1, 2]);
octave:31> taps
taps =

  -6.26930101730182e-04
  -1.03354450635036e-01
  -9.81576747564301e-03
   3.17271686090449e-01
   5.11409425599933e-01
   3.17271686090449e-01
  -9.81576747564301e-03
  -1.03354450635036e-01
  -6.26930101730182e-04
```